jetty-io/src/main/java/org/eclipse/jetty/io/ByteBufferAccumulator.java

```diff
@@ -0,0 +1,138 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.eclipse.jetty.util.BufferUtil;
+
+public class ByteBufferAccumulator implements AutoCloseable
+{
+    private final List<ByteBuffer> _buffers = new ArrayList<>();
+    private final ByteBufferPool _bufferPool;
+
+    public ByteBufferAccumulator()
+    {
+        this(null);
+    }
+
+    public ByteBufferAccumulator(ByteBufferPool bufferPool)
+    {
+        _bufferPool = (bufferPool == null) ? new NullByteBufferPool() : bufferPool;
+    }
+
+    public int getLength()
+    {
+        int length = 0;
+        for (ByteBuffer buffer : _buffers)
+            length += buffer.remaining();
+        return length;
+    }
+
+    public ByteBuffer ensureBuffer(int minSize, int minAllocationSize)
+    {
+        ByteBuffer buffer = _buffers.isEmpty() ? BufferUtil.EMPTY_BUFFER : _buffers.get(_buffers.size() - 1);
+        if (BufferUtil.space(buffer) <= minSize)
+        {
+            buffer = _bufferPool.acquire(minAllocationSize, false);
+            _buffers.add(buffer);
+        }
+
+        return buffer;
+    }
+
+    public void copyBytes(byte[] buf, int offset, int length)
+    {
+        copyBuffer(BufferUtil.toBuffer(buf, offset, length));
+    }
+
+    public void copyBuffer(ByteBuffer buffer)
+    {
+        while (buffer.hasRemaining())
+        {
+            ByteBuffer b = ensureBuffer(0, buffer.remaining());
+            int pos = BufferUtil.flipToFill(b);
+            BufferUtil.put(buffer, b);
+            BufferUtil.flipToFlush(b, pos);
+        }
+    }
+
+    public ByteBuffer takeByteBuffer()
+    {
+        int length = getLength();
+        ByteBuffer combinedBuffer = _bufferPool.acquire(length, false);
+        for (ByteBuffer buffer : _buffers)
+        {
+            combinedBuffer.put(buffer);
+        }
+        return combinedBuffer;
+    }
```

Comment:
```suggestion
    public ByteBuffer takeByteBuffer()
    {
        int length = getLength();
        switch(length)
        {
            case 0:
                return null; // TODO or empty buffer from pool?
            case 1:
                return _buffers.remove(0);
            default:
                ByteBuffer combinedBuffer = _bufferPool.acquire(length, false);
                for (ByteBuffer buffer : _buffers)
                {
                    combinedBuffer.put(buffer);
                    _bufferPool.release(buffer);
                }
                _buffers.clear();
                return combinedBuffer;
        }
    }
```

Reply:
I don't really like the idea of returning null from here, it will introduce null checks everywhere this is used.
How can we get empty buffer from pool, will `_bufferPool.aquire(0, false)` work, because that's what we are currently doing for 0 length.

Reply:
This also makes it so you can only call `takeBuffer()` once, but you can call `toBuffer()` multiple times and get the same buffer.

Why wouldn't we just release these buffers on `close()` like we would do in other cases?

Reply:
Because the idea of `takeBuffer` is that you take the buffer and all it's data, hence it can't be called again.
The intent is that once you call `takeBuffer` you will pass the resulting buffer along some further chain of events and then only release it once it is free.
freeing the buffer on close makes no sense for most uses, as we want to have a accumulator that lives as long as the connection, during which time it may create many buffers that are taken and passed on.

Happy not to return null.